### PR TITLE
Email/get skeleton

### PR DESCRIPTION
**Pull request by chibenwa**
This method does the skeleton work for Email/get, so far allowing only to get the `size` property of an Email.

Property filtering is implemented.

CreationId handling is out of scope.

### CODE REVIEW

File: server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/method/EmailGetMethod.scala [hunk 1]
```diff
@@ -69,17 +69,36 @@ class EmailGetMethod @Inject() (serializer: Serializer,
...
         .onErrorResume({
           case e: IllegalArgumentException => SMono.just(Invocation.error(ErrorCode.InvalidArguments, e.getMessage, invocation.methodCallId))
           case e: Throwable => SMono.raiseError(e)
         })
         .map(invocationResult => (invocationResult, processingContext)))
 
+  private def computeResponseInvocation(request: EmailGetRequest, invocation: Invocation, mailboxSession: MailboxSession): SMono[Invocation] =
+    validateProperties(request)
+        .fold(
+          e => SMono.raiseError(e),
```
Comment (Arsnael) [inline server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/method/EmailGetMethod.scala:82]: no need to return to the line here?

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/EmailGetMethodContract.scala [hunk 2]
```diff
@@ -450,6 +450,171 @@ trait EmailGetMethodContract {
...
+         |    "methodResponses": [[
+         |            "error",
+         |            {
+         |                "type": "invalidArguments",
+         |                "description": "The following properties [invalid] do not exist."
+         |            },
+         |            "c1"
+         |        ]]
+         |}""".stripMargin)
+  }
```
Comment (Arsnael) [inline server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/EmailGetMethodContract.scala:616]: How about id should be returned even when properties is empty `[]` ?

**Review by Arsnael — COMMENTED** [review]
I would say serialization unit tests regarding Email and EmailGet (Request and Response) are missing as well?

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/EmailGetMethodContract.scala [hunk 3]
```diff
@@ -0,0 +1,496 @@
...
+
+    requestSpecification = baseRequestSpecBuilder(server)
+      .setAuth(authScheme(UserCredential(BOB, BOB_PASSWORD)))
+      .build
+  }
+
+  def randomMessageId: MessageId
+
+  @Test
+  def idsShouldBeMandatory(): Unit = {
```
[the remark this answers is not in the document]
Comment (chibenwa) [reply server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/EmailGetMethodContract.scala:55]: There is no mention that the ids in the get object can be null for Email (contrary to mailbox)

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/EmailGetMethodContract.scala [hunk 4]
```diff
@@ -0,0 +1,496 @@
...
+         |            {
+         |                "accountId": "29883977c13473ae7cb7678ef767cbfbaffc8a44a6e463d971d23a65c1dc4af6",
+         |                "state": "000001",
+         |                "list": [],
+         |                "notFound": ["${messageId.serialize()}"]
+         |            },
+         |            "c1"
+         |        ]]
+         |}""".stripMargin)
+  }
```
[the remark this answers is not in the document]
Comment (chibenwa) [reply server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/EmailGetMethodContract.scala:660]: > capabilities core and mail should be present?

Handled as part of #3745 ?

> - message in a shared mailbox?
> - I should not be able to fetch the mail of an other user?

Well that's a nice idea, I'll contribute it tomorrow.

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/EmailGetMethodContract.scala (hunk 3 above)
[the remark this answers is not in the document]
Comment (Arsnael) [reply server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/EmailGetMethodContract.scala:55]: Just checked, you are totally right :)

File: server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/method/EmailGetMethod.scala (hunk 1 above)
Comment (chibenwa) [reply server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/method/EmailGetMethod.scala:82]: I did prefer having the two case handled at the same level, it makes it visually more clear.

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/EmailGetMethodContract.scala (hunk 2 above)
Comment (chibenwa) [reply server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/EmailGetMethodContract.scala:616]: :+1:

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/EmailGetMethodContract.scala (hunk 4 above)
[the remark this answers is not in the document]
Comment (Arsnael) [reply server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/EmailGetMethodContract.scala:660]: > > capabilities core and mail should be present?
> 
> Handled as part of #3745 ?

I would be careful on that... It depends which PR gets approved and merged first^^'

**Comment by chibenwa** [comment]
`Exception in thread "Thread-1723" java.lang.RuntimeException: java.io.IOException: Failed to bind to /0.0.0.0:41133`

**Comment by chibenwa** [comment]
test this please

**Comment by chibenwa** [comment]
> I would say serialization unit tests regarding Email and EmailGet (Request and Response) are missing as well?

Overkill to me. It adds little value as it is already covered by integration tests.

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/EmailGetMethodContract.scala [hunk 5]
```diff
@@ -714,4 +715,176 @@ trait EmailGetMethodContract {
...
+         |                    }],
+         |                "notFound": []
+         |            },
+         |            "c1"
+         |        ]]
+         |}""".stripMargin)
+  }
+
+  @Test
+  def getShouldReturnnotFoundWhenDeletedUserMissesReadRight(server: GuiceJamesServer): Unit = {
```
Comment (Arsnael) [inline server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/EmailGetMethodContract.scala:835]:
```suggestion
  def getShouldReturnNotFoundWhenDeletedUserMissesReadRight(server: GuiceJamesServer): Unit = {
```

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/EmailGetMethodContract.scala (hunk 4 above)
[the remark this answers is not in the document]
Comment (chibenwa) [reply server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/EmailGetMethodContract.scala:717]: No need to handle the same thing twice.

File: server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/mail/Email.scala [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+import org.apache.james.mailbox.model.{MessageId, MessageResult}
+import org.slf4j.{Logger, LoggerFactory}
+
+object Email {
+  private val logger: Logger = LoggerFactory.getLogger(classOf[Email])
+
+  type UnparsedEmailIdConstraint = NonEmpty
+  type UnparsedEmailId = String Refined UnparsedEmailIdConstraint
+
+  def asUnparsed(messageId: MessageId): UnparsedEmailId = refined.refineV[UnparsedEmailIdConstraint](messageId.serialize()) match {
```
Comment (remk) [inline server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/mail/Email.scala:43]: Not a fan of lying on the method signature as it should be a Try[UnparsedEmailIdConstraint]

**Review by remk — APPROVED** [review]

**Comment by chibenwa** [comment]
test this please

File: server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/mail/Email.scala [hunk 7]
```diff
@@ -40,9 +42,10 @@ object Email {
   val allowedProperties: Properties = Properties(Set("id", "size"))
   val idProperty: Properties = Properties(Set("id"))
 
-  def asUnparsed(messageId: MessageId): UnparsedEmailId = refined.refineV[UnparsedEmailIdConstraint](messageId.serialize()) match {
-    case Left(e) => throw new IllegalArgumentException(e)
-    case scala.Right(value) => value
+  def asUnparsed(messageId: MessageId): Try[UnparsedEmailId] =
+    refined.refineV[UnparsedEmailIdConstraint](messageId.serialize()) match {
+    case Left(e) => Failure(new IllegalArgumentException(e))
+    case scala.Right(value) => Success(value)
```
Comment (Arsnael) [inline server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/mail/Email.scala:48]: indent?

**Review by Arsnael — APPROVED** [review]
Don't forget to have a JIRA ticket associated with this PR :)

**Review by remk — APPROVED** [review]

**Comment by Arsnael** [comment]
```
[323df4f38fb6694aed60948ad87988ebe5717c57] [ERROR] selectThenFetchWithExistingMessages  Time elapsed: 0.875 s  <<< ERROR!
[323df4f38fb6694aed60948ad87988ebe5717c57] org.apache.james.mpt.protocol.ProtocolSession$InvalidServerResponseException: 
[323df4f38fb6694aed60948ad87988ebe5717c57] 
[323df4f38fb6694aed60948ad87988ebe5717c57] Location: /org/apache/james/imap/scripts/SelectThenFetchWithExistingMessages.test:51
[323df4f38fb6694aed60948ad87988ebe5717c57] LastClientMsg: A06 DELETE selectTest
[323df4f38fb6694aed60948ad87988ebe5717c57] Expected: 'A06 OK DELETE completed\.'
[323df4f38fb6694aed60948ad87988ebe5717c57] Actual   : 'A06 NO DELETE processing failed.'
[323df4f38fb6694aed60948ad87988ebe5717c57] 
[323df4f38fb6694aed60948ad87988ebe5717c57] [INFO] 
[323df4f38fb6694aed60948ad87988ebe5717c57] [INFO] Results:
[323df4f38fb6694aed60948ad87988ebe5717c57] [INFO] 
[323df4f38fb6694aed60948ad87988ebe5717c57] [ERROR] Errors: 
[323df4f38fb6694aed60948ad87988ebe5717c57] [ERROR]   JamesDeploymentValidationTest>DeploymentValidation.selectThenFetchWithExistingMessages:86 » InvalidServerResponse
```
deployment tests...

File: server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/mail/Email.scala (hunk 7 above)
Comment (Arsnael) [reply server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/mail/Email.scala]: Fixed

File: server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/method/EmailGetMethod.scala (hunk 1 above)
Comment (LanKhuat) [inline server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/method/EmailGetMethod.scala:89]: Indent?

**Review by LanKhuat — APPROVED** [review]

File: server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/method/EmailGetMethod.scala (hunk 1 above)
Comment (chibenwa) [reply server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/method/EmailGetMethod.scala:89]: I prefer having the two parameters of the fold on the same indent level, it makes things more readable...

**Review by remk — APPROVED** [review]

**Comment by Arsnael** [comment]
Merged

**Pull request closed without merging**